### ISSUE TITLE
Getting error 'unicode' object has no attribute 'get'

### ISSUE THREAD

**Issue by atrauzzi** [Azure/azure-cli#2630]
When I try to run this command: https://github.com/atrauzzi/anecdote-azure/blob/master/bin/deploy-functions#L68-L76 , I end up getting this error:

```
'unicode' object has no attribute 'get'
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/azure/cli/main.py", line 37, in main
    cmd_result = APPLICATION.execute(args)
  File "/usr/local/lib/python2.7/dist-packages/azure/cli/core/application.py", line 157, in execute
    result = expanded_arg.func(params)
  File "/usr/local/lib/python2.7/dist-packages/azure/cli/core/commands/arm.py", line 267, in handler
    instance = getter(client, **getterargs) if client else getter(**getterargs)
  File "/usr/local/lib/python2.7/dist-packages/azure/mgmt/resource/resources/operations/resources_operations.py", line 520, in get
    exp = CloudError(response)
  File "/usr/local/lib/python2.7/dist-packages/msrestazure/azure_exceptions.py", line 136, in __init__
    self._build_error_data(response)
  File "/usr/local/lib/python2.7/dist-packages/msrestazure/azure_exceptions.py", line 156, in _build_error_data
    data = data.get('error', data)
AttributeError: 'unicode' object has no attribute 'get'
```

**Comment by yugangw-msft** [comment]
Likely you ran into an error, and the error is not in json format which is a well known webapp service bug.
Suggest run the command manually with '--debug' and let me know what you get.

**Comment by atrauzzi** [comment]
```
msrest.http_logger : Response status: 404
msrest.http_logger : Response headers:
msrest.http_logger :     'Cache-Control': 'no-cache'
msrest.http_logger :     'Pragma': 'no-cache'
msrest.http_logger :     'Content-Length': '23'
msrest.http_logger :     'Content-Type': 'application/json; charset=utf-8'
msrest.http_logger :     'Expires': '-1'
msrest.http_logger :     'Strict-Transport-Security': 'max-age=31536000; includeSubDomains'
msrest.http_logger :     'Set-Cookie': 'ARRAffinity=REDACTED;Path=/;Domain=atrauzzi-blog-engine.scm.azurewebsites.net'
msrest.http_logger :     'Server': 'Microsoft-IIS/8.0, Microsoft-IIS/8.0'
msrest.http_logger :     'X-AspNet-Version': '4.0.30319'
msrest.http_logger :     'X-Powered-By': 'ASP.NET, ASP.NET'
msrest.http_logger :     'x-ms-ratelimit-remaining-subscription-reads': '14999'
msrest.http_logger :     'x-ms-request-id': 'REDACTED'
msrest.http_logger :     'x-ms-correlation-request-id': 'REDACTED'
msrest.http_logger :     'x-ms-routing-request-id': 'NORTHCENTRALUS:20170329T011102Z:REDACTED'
msrest.http_logger :     'Date': 'Wed, 29 Mar 2017 01:11:01 GMT'
msrest.http_logger : Response content:
msrest.http_logger : "Deployment not found."
```

Okay, so it's saying deployment not found, although I've been following [these instructions](http://stackoverflow.com/a/42895510/128991).  Not sure how to correct in this case.

**Comment by yugangw-msft** [comment]
//cc: @davidebbo

**Comment by davidebbo** [comment]
It could mean that the package url you gave it is 404. Try running the command manually with hard coded value to isolate from any possible issues in your batch file token logic.

**Comment by atrauzzi** [comment]
@davidebbo - Here's the command that ends up being generated:

```
az resource update \
    --api-version 2015-08-01 \
    --namespace Microsoft.Web \
    --name MSDeploy \
    --resource-type Extensions \
    --resource-group atrauzzi-blog \
    --parent sites/atrauzzi-blog-engine \
    --set "properties.packageUri=https://atrauzziblog.blob.core.windows.net/deployment/functions.zip"
```

Feel free to try that URI yourself, nothing sensitive in there.  But I'm able to access it.

**Comment by davidebbo** [comment]
I'm wondering if something changes in the CLI. All this command is trying to do is a PUT request on the msdeploy resource. But for some reason it's doing a get before the put, and that returns a 404. @yugangw-msft do you think it could be a cli regression? Nothing I can think of has changed on the App Service side.

**Comment by davidebbo** [comment]
Ok, so I have a theory: if you deployed anything before to that app using msdeploy, the GET works, and allows the command to run. To test the theory, try this:

- In Kudu Console, go to `D:\home\LogFiles\SiteExtensions\MSDeploy` (create folders as needed)
- In there, create a file called `appManagerStatus.xml`, containing:

```xml
<?xml version="1.0" encoding="utf-8"?>
<deployment>
</deployment>
```

Then try running your az command again.

**Comment by atrauzzi** [comment]
Giving this a shot, will edit/update this comment after...Stand by!

**Comment by atrauzzi** [comment]
@davidebbo - Well, added the file and re-ran the command.  The only thing I can say is now it's just sitting there doing nothing.  Has been at it for about 10 or so minutes.

So either it's working extremely slowly, or I've somehow slipped between the cracks again.

**Comment by davidebbo** [comment]
Probably just slow, expanding your Node modules. Give it a bit more time :)

**Comment by atrauzzi** [comment]
Hmm, nope.

Got this error: 

```
Error occurred in request., ConnectionError: HTTPSConnectionPool(host='management.azure.com', port=443): Max retries exceeded with url: /subscriptions/REDACTED/resourcegroups/atrauzzi-blog/providers/Microsoft.Web/sites/atrauzzi-blog-engine/Extensions/MSDeploy/status?api-version=2015-08-01 (Caused by ProtocolError('Connection aborted.', Ba
dStatusLine("''",)))
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/azure/cli/main.py", line 37, in main
    cmd_result = APPLICATION.execute(args)
  File "/usr/local/lib/python2.7/dist-packages/azure/cli/core/application.py", line 157, in execute
    result = expanded_arg.func(params)
  File "/usr/local/lib/python2.7/dist-packages/azure/cli/core/commands/arm.py", line 318, in handler
    result = opres.result() if isinstance(opres, AzureOperationPoller) else opres
  File "/usr/local/lib/python2.7/dist-packages/msrestazure/azure_operation.py", line 473, in result
    self.wait(timeout)
  File "/usr/local/lib/python2.7/dist-packages/msrestazure/azure_operation.py", line 486, in wait
    raise self._exception
ClientRequestError: Error occurred in request., ConnectionError: HTTPSConnectionPool(host='management.azure.com', port=443): Max retries exceeded with url: /subscriptions/REDACTED/resourcegroups/atrauzzi-blog/providers/Microsoft.Web/sites/atrauzzi-blog-engine/Extensions/MSDeploy/status?api-version=2015-08-01 (Caused by ProtocolError('Conn
ection aborted.', BadStatusLine("''",)))
Starting App Service
Error occurred in request., ConnectionError: HTTPSConnectionPool(host='management.azure.com', port=443): Max retries exceeded with url: /subscriptions/REDACTED/resourceGroups/atrauzzi-blog/providers/Microsoft.Web/sites/atrauzzi-blog-engine/start?api-version=2016-08-01 (Caused by ProtocolError('Connection aborted.', BadStatusLine("''",)))
```

**Comment by davidebbo** [comment]
I think there are some unrelated issues here, and it's best not to try to tackle them all at once. Start with a clean wwwroot (use Kudu), and deploy a trivial zip with a few files, and see that works consistently. If it does, then that settles the original issue that started this thread (which is still something we probably need to fix in the CLI, as it shouldn't be doing a GET when we don't want it).

Then if that works, we can look at what's happening with the bigger file, which is likely unrelated.

**Comment by yugangw-msft** [comment]
> need to fix in the CLI, as it shouldn't be doing a GET when we don't want it

@davidebbo, the generic resource update assumes the resource exists, so it does always do a `GET` first.
We can consider to handle the case of `404`

**Comment by davidebbo** [comment]
@yugangw-msft note that when deploying from an ARM template, it just does a PUT and no GET.

And in PowerShell, you can choose between Set-AzureRmResource and New-AzureRmResource depending on whether you want a GET first or a just a PUT. Generally, I think there are a number of scenarios where the GET is undesirable.

Could there maybe be some kind of  `--no-get` option? Either that or an alternate `az resource create`.

**Comment by yugangw-msft** [comment]
@davidebbo, i lean forward to letting `az resource create` handle it

**Comment by atrauzzi** [comment]
@davidebbo - Just a quick update, I'll see about trying your suggestion above this evening.

**Comment by atrauzzi** [comment]
@davidebbo - Well whaddaya know!

```
{
  "id": "/subscriptions/2fbf65ec-7fa9-470d-819a-2838b8bc66ad/resourcegroups/atrauzzi-blog/providers/Microsoft.Web/sites/
atrauzzi-blog-engine/Extensions/MSDeploy",
  "identity": null,
  "kind": null,
  "location": "East US 2",
  "managedBy": null,
  "name": "MSDeploy",
  "plan": null,
  "properties": {
    "complete": true,
    "deployer": "58859F63FA3177F6EC558D4C409DA451A7C69A7C",
    "endTime": "2017-03-30T23:58:58.1099248+00:00",
    "provisioningState": "succeeded",
    "startTime": "2017-03-30T23:58:57.5567518+00:00"
  },
  "sku": null,
  "tags": null,
  "type": "Microsoft.Web/sites/extensions"
}
```

![image](https://cloud.githubusercontent.com/assets/115199/24530993/58ecab54-157b-11e7-9709-cd7757e2a57d.png)

The test file I used is there!  Just about to try another deploy of the regular suite, but I suspect it'll die again.  Are we thinking this is a timeout issue from `az`?

**Comment by atrauzzi** [comment]
@davidebbo - Well hey, that deployment worked!

![image](https://cloud.githubusercontent.com/assets/115199/24531297/83184404-157d-11e7-8c10-8073d31c095c.png)

So I guess the thing here is that I did reduce the size of the package overall by sharing the `node_modules` dir between functions.

My only concern beyond what we've already identified is that there perhaps appears to be some risk of deployments timing out?
